Change Flash retry logic to use backoff delays

AndrejMitrovic opened the pull request:
This also includes a bunch of fixes to the internals.

But let's see what the CI says.

Fixes #1890

AndrejMitrovic committed:
Fix styling

AndrejMitrovic committed:
Flash: Pass FlashConfig to UpdateSigner

It will be needed to use proper retry
delays and max retries.

AndrejMitrovic committed:
Flash: Add error return for UpdateSigner

This will allow the calling code to notify
on failure rather than keep UpdateSigner
stuck forever trying to get the signatures.

AndrejMitrovic committed:
Implement truncating exponential backoff algorithm

This implementation is based on Amazon's description
on https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

It will be used in the Flash layer, and potentially other
parts of code which require using retry delays.

AndrejMitrovic committed:
Flash: Replace max retries with `max_retry_time`

This is a much more reliable measure since it's easier
for the user to set a maximum retry time rather than try
to manually calculate it via multiplication.

Plus manual calculation is not possible if randomization
is added to the delays later.

AndrejMitrovic committed:
Flash: Add max_retry_delay option

Together with the `max_payment_retries` option,
it will be used together with the truncated exponential
backoff retrying algorithm.

AndrejMitrovic committed:
Flash: Integrate backoff retry algorithm

AndrejMitrovic committed:
Flash: Add retry_multiplier config option

Also used for the backoff algorithm.

AndrejMitrovic committed:
Flash: Add better logging for mismatching sequences

Fixes #1890

AndrejMitrovic committed:
Flash: Return error code in confirmChannelUpdate()

The calling peer must retry calling this method
until the receiver accepts it or until enough
retries have been attempted.

AndrejMitrovic commented:
> submodules/agora/source/agora/flash/UpdateSigner.d(35,8): Error: module `Backoff` is in file 'agora/utils/Backoff.d' which cannot be read

So every time we add a new module we first have to update Faucet's `dub.json` file and add that file there, even without knowing whether the PR will be accepted. This is ridiculous.

omerfirmak commented on source/agora/flash/Channel.d:
```diff
@@ -1157,13 +1185,24 @@ LOuter: while (1)
     {
         const old_balance = this.cur_balance;
 
+        const old_seq_id = this.cur_seq_id;
```
`collectSignatures` already takes `update.seq_id` as an argument. Could we wait until `collectSignatures` to return to update the `cur_seq_id`?

AndrejMitrovic replied:
Sure, good idea.

omerfirmak commented on source/agora/flash/Node.d:
```diff
@@ -853,10 +853,11 @@ public abstract class FlashNode : ControlFlashAPI
         {
-            assert(!error); // Payee should never fail to receive
-            channel.learnSecrets([*secret], [], this.last_block_height);
+            if (!error)
```
I can't see how, is there an example you can give? AFAICS, `onPaymentComplete` is never called with an error for incoming payments anyways.

AndrejMitrovic replied:
The signing process can fail. I added a notification for failed incoming payments, but now that I think about it maybe it should be a separate callback.

I'll rework it a bit.

AndrejMitrovic commented:
Thanks for the tips @omerfirmak. I removed the incoming payment change and fixed up the sequences.

AndrejMitrovic commented:
It looks like https://github.com/bosagora/faucet/pull/56 isn't ready yet so we'll have to do this in a funky way by juggling submodules..

AndrejMitrovic commented:
Edit: s/submodules/docker images

AndrejMitrovic commented:
It might be easier to just merge it first, then fix faucet to add the new import, and then update agora again. Let's wait for green from the CI.

AndrejMitrovic committed:
Flash: Fix comparison of ChannelUpdate

Signatures should not be compared when deciding
on whether a channel update is new or not.

AndrejMitrovic committed:
Flash: Fix missing printouts

AndrejMitrovic commented:
Merging this, and I'll fix up the faucet version soon.
